### ISSUE TITLE
torch.bernoulli() randomly returns "1" for 0 inputs on CPU

### ISSUE THREAD

**Issue by gletarte** [pytorch/pytorch#26807]
## 🐛 Bug
`torch.bernoulli()` randomly returns "1" for 0 inputs, even if the probability of drawing "1" is 0.
This bug occurs only on CPU.

## To Reproduce

Run the following snippet:

```python
import torch
a = torch.zeros(10000, 10000) # probability of drawing "1" is 0
print((torch.bernoulli(a) == 1).sum())
```
It yields a value between 3-11, which is the number of "1" drawn on 100 million drawings.

## Expected behavior

`torch.bernoulli(input)` should always return "0" for every 0 in input. 

## Environment

PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Arch Linux
GCC version: (GCC) 9.1.0
CMake version: version 3.15.3

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.14.3
[pip] numpydoc==0.8.0
[pip] torch==1.2.0
[pip] torchvision==0.4.0a0+6b959ee
[conda] blas                      1.0                         mkl  
[conda] mkl                       2018.0.2                      1  
[conda] mkl-service               1.1behavior.2            py36h17a0993_4  
[conda] mkl_fft                   1.0.1            py36h3010b51_0  
[conda] mkl_random                1.0.1            py36h629b387_0  
[conda] pytorch                   1.2.0           py3.6_cuda10.0.130_cudnn7.6.2_0    pytorch
[conda] torch                     0.4.1                    pypi_0    pypi
[conda] torchvision               0.4.0                py36_cu100    pytorch

## Additional context

This bug does not occur if we specify to use type `torch.float64`: 

```python
import torch
a = torch.zeros(10000, 10000, dtype=torch.float64) # probability of drawing "1" is 0
print((torch.bernoulli(a) == 1).sum())
```
which leads to believe that a casting from `torch.float32` to `torch.float64` in `torch.bernoulli()` sometimes converts some zero values to small values, yielding the unwanted behavior.

Finally, this bug does not happen on GPU.

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

**Comment by freud14** [comment]
I just tested this on version 0.4.1 on CPU and the bug is not present. However, on version 1.2, it is.

**Comment by Lalaland** [comment]
This is due to https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/DistributionsHelper.h#L171. Note how it uses a <= when it should use a <.

The real question here is whether or not we are allowed to change that because it will result in a change of behavior for a fixed seed.

**Comment by Balandat** [comment]
> The real question here is whether or not we are allowed to change that because it will result in a change in behavior for a fixed seed.

That should be extremely rare though, no? Like it would *almost surely* never happen (sweeping precision issues under the rug here). 

An alternative could be to just special case the cases `p==0.0` and `p=1.0` and return deterministic zeros/ones.

**Comment by Lalaland** [comment]
This is now fixed as of the merge of #26864.

**Comment by vincentqb** [comment]
Great! I'll close this then. Please feel free to re-open if the problem persists.
